File: examples/js/02_node_main_api/students.ts
// Copyright 2022 Touca, Inc. Subject to Apache-2.0 License.

import { touca } from '@touca/node';

interface Course {
  name: string;
  grade: number;
}

interface Student {
  username: string;
  fullname: string;
  dob: Date;
  gpa: number;
}

const students = [
  {
    username: 'alice',
    fullname: 'Alice Anderson',
    dob: new Date(2006, 3, 1),
    courses: [
      { name: 'computers', grade: 3.8 },
      { name: 'math', grade: 4.0 }
    ]
  },
  {
    username: 'bob',
    fullname: 'Bob Brown',
    dob: new Date(1996, 6, 30),
    courses: [
      { name: 'history', grade: 3.9 },
      { name: 'english', grade: 3.7 }
    ]
  },
  {
    username: 'charlie',
    fullname: 'Charlie Clark',
    dob: new Date(2003, 9, 19),
    courses: [
      { name: 'computers', grade: 3.7 },
      { name: 'math', grade: 2.9 }
    ]
  }
];

function calculate_gpa(courses: Course[]): number {
  for (const course of courses) {
    touca.add_array_element('courses', course);
    touca.add_hit_count('number of courses');
  }
  return courses.length
    ? courses.reduce((sum, v) => sum + v.grade, 0) / courses.length
    : 0.0;
}

export async function find_student(username: string): Promise<Student> {
  await new Promise((v) => setTimeout(v, 100));
  const data = students.find((v) => v.username === username);
  if (!data) {
    throw new Error(`no student found for username: ${username}`);
  }
  const { courses, ...student } = data;
  return { ...student, gpa: calculate_gpa(courses) };
}

File: examples/python/02_python_main_api/students.py
# Copyright 2022 Touca, Inc. Subject to Apache-2.0 License.

from dataclasses import dataclass
from datetime import date
from random import random
from time import sleep
from typing import List

import touca


@dataclass
class Course:
    name: str
    grade: float


@dataclass
class Student:
    username: str
    fullname: str
    dob: date
    gpa: float


students = [
    (
        "alice",
        "Alice Anderson",
        date(2006, 3, 1),
        [Course("computers", 3.8), Course("math", 4.0)],
    ),
    (
        "bob",
        "Bob Brown",
        date(1996, 6, 30),
        [Course("history", 3.9), Course("english", 3.7)],
    ),
    (
        "charlie",
        "Charlie Clark",
        date(2003, 9, 19),
        [Course("computers", 3.7), Course("math", 2.9)],
    ),
]


def calculate_gpa(courses: List[Course]):
    for course in courses:
        touca.add_array_element("courses", course)
        touca.add_hit_count("number of courses")
    return sum(k.grade for k in courses) / len(courses) if courses else 0


def find_student(username: str) -> Student:
    sleep(0.2 + random() * 0.05)
    data = next((k for k in students if k[0] == username), None)
    if not data:
        raise ValueError(f"no student found for username: {username}")
    return Student(data[0], data[1], data[2], calculate_gpa(data[3]))
